---
title: "HKU workshop"
author: "Sydney Precision Bioinformatics Group"
date: "03/05/2019"
output: html_document
---

This is the index page. 


# The Single Cell Plus Workshop

This workshop is part of a joint research initiative between Hong Kong University and the University of Sydney. 


In this workshop we will present thre main topics: 

  1. Quality control of sc-RNASeq data,
  1. The `scMerge` package for integrating multiple sc-RNASeq data, 
  1. Some possible downstream analysis after `scMerge`. 


At the end of this workshop you should have some basic understanding of scRNASeq data and how to deal with them. 


# Prerequisites

Ideally, you should be somewhat familiar with `R`. However, if you haven't used `R` before, don't worry! This is not a programming workshop, you can still pick up important practical skills by running the codes we provided. 

  + `R 3.6` from https://cloud.r-project.org/ 
  + `RStudio` from https://www.rstudio.com/products/rstudio/download/
  

# Installation 

```{r, eval = FALSE}
install.packages(c("BiocManager", "devtools"))
BiocManager::install(c("scMerge", "scran", "scater", "monocle", "Rtsne", "DropletUtils"))
devtools::install_github("SydneyBioX/scdney")
```

## For Mac users

If you could not run `devtools::install_github("SydneyBioX/scdney")`, then please try the following code instead. 

```{r, eval = FALSE}
install.packages("https://github.com/kevinwang09/SingleCellPlus/raw/master/scdney_mac_binary/scdney_0.1.2.tgz", repos = NULL)
```

