---
title: "HKU workshop"
author: "Sydney Precision Bioinformatics Group"
date: "18 June 2019"
output: html_document
---

# The Single Cell Plus Workshop

This workshop is part of a joint research initiative between Hong Kong University and the University of Sydney.


# Aim of this workshop

In this workshop we will present thre main topics that are critical to single-cell analysis. We will mainly be using 2 mouse liver data for analysis. 

  1. Quality control of sc-RNASeq data,
  1. The `scMerge` package for integrating multiple sc-RNASeq data, 
  1. Some possible downstream analysis after `scMerge`. 


At the end of this workshop you should have some basic understanding of scRNASeq data and some key computational and statistical challenges of this type of data. 


# Prerequisites

Ideally, you should be somewhat familiar with `R`. However, if you haven't used `R` before, don't worry! This is not a programming workshop, you can still pick up important practical skills by running the codes we provided. 

  + `R 3.6` from https://cloud.r-project.org/ 
  + `RStudio` from https://www.rstudio.com/products/rstudio/download/
  

# Installation 

You should be able to run the codes below in `R` to install all the packages needed in this workshop. If you encounter any problems. Please let one of the instructors know. 

```{r install, eval = FALSE}
install.packages("BiocManager")
BiocManager::install("scMerge", version = "devel")
BiocManager::install(c("devtools", "DropletUtils", "edgeR",
                       "ggpubr", "MAST", 
                       "mclust", "monocle", 
                       "plyr", "Rtsne", 
                       "scales", "scater", 
                       "scran", "tidyverse", "viridis",
                       "DescTools", "amap", "doParallel", "ggridges", "lme4", "mice", "methods", "caret", "randomForest", "clusteval"
                       ))
devtools::install_github("SydneyBioX/scdney")
```

## For Mac users

If you could not run the last line of the above instructions (`devtools::install_github("SydneyBioX/scdney")`), then please try the following code instead. 

```{r mac install, eval = FALSE}
install.packages("https://github.com/kevinwang09/SingleCellPlus/raw/master/scdney_mac_binary/scdney_0.1.2.tgz", repos = NULL)
```

# Reference

**Methodologies**

1. [scMerge leverages factor analysis, stable expression, and pseudoreplication to merge multiple single-cell RNA-seq datasets]((http://www.pnas.org/lookup/doi/10.1073/pnas.1820006116)) (2019). Yingxin Lin, Shila Ghazanfar, Kevin Y.X. Wang, Johann A. Gagnon-Bartsch, Kitty K. Lo, Xianbin Su, Ze-Guang Han, John T. Ormerod, Terence P. Speed, Pengyi Yang, Jean Y. H. Yang.

1. [Evaluating stably expressed genes in single cells](https://www.biorxiv.org/content/biorxiv/early/2018/11/22/229815.full.pdf) (Under review). Yingxin Lin, Shila Ghazanfar, Dario Strbenac, Andy Wang, Ellis Patrick, Dave Lin, Terence Speed, Jean Y. H. Yang, Pengyi Yang.

1. [Impact of similarity metrics on single-cell RNA-seq data clustering](https://doi.org/10.1093/bib/bby076) (2018). Taiyun Kim, Irene Rui Chen, Yingxin Lin, Andy Y. Y. Wang, Jean Y. H. Yang, Pengyi Yang.


**Data**

1. [Single-cell RNA-Seq analysis reveals dynamic trajectories during mouse liver development](https://doi.org/10.1186/s12864-017-4342-x) (2017). Xianbin Su, Yi Shi, Xin Zou, Zhao-Ning Lu, Gangcai Xie, Jean Y. H. Yang, Chong-Chao Wu, Xiao-Fang Cui, Kun-Yan He, Qing Luo, Yu-Lan Qu, Na Wang, Lan Wang, Ze-Guang Han.

1. [A single‐cell transcriptomic analysis reveals precise pathways and regulatory mechanisms underlying hepatoblast differentiation](https://doi.org/10.1002/hep.29353) (2017). Li Yang, Wei‐Hua Wang, Wei‐Lin Qi, Zhen Guo, Erfei Bi, Cheng‐Ran Xu. 



# Session Info
```{r}
Sys.time()
sessionInfo()
```

