---
title: "HKU workshop"
author: "Sydney Precision Bioinformatics Group"
date: "18 June 2019"
output: html_document
---

# The Single Cell Plus Workshop

This workshop is part of a joint research initiative between Hong Kong University and the University of Sydney.


# Aim of this workshop

In this workshop we will present thre main topics that are critical to single-cell analysis. We will mainly be using 2 mouse liver data for analysis. 

  1. Quality control of sc-RNASeq data,
  1. The `scMerge` package for integrating multiple sc-RNASeq data, 
  1. Some possible downstream analysis after `scMerge`. 


At the end of this workshop you should have some basic understanding of scRNASeq data and some key computational and statistical challenges of this type of data. 


# Prerequisites

Ideally, you should be somewhat familiar with `R`. However, if you haven't used `R` before, don't worry! This is not a programming workshop, you can still pick up important practical skills by running the codes we provided. 

  + `R 3.6` from https://cloud.r-project.org/ 
  + `RStudio` from https://www.rstudio.com/products/rstudio/download/
  

# Installation 

You should be able to run the codes below in `R` to install all the packages needed in this workshop. If you encounter any problems. Please let one of the instructors know. 

```{r install, eval = FALSE}
install.packages("BiocManager")

BiocManager::install(c("scMerge", "devtools",
                       "DropletUtils", "edgeR",
                       "ggpubr", "MAST", 
                       "mclust", "monocle", 
                       "plyr", "Rtsne", 
                       "scales", "scater", 
                       "scran", "tidyverse", "viridis"))

devtools::install_github("SydneyBioX/scdney")
```

## For Mac users

If you could not run the last line of the above instructions (`devtools::install_github("SydneyBioX/scdney")`), then please try the following code instead. 

```{r mac install, eval = FALSE}
install.packages("https://github.com/kevinwang09/SingleCellPlus/raw/master/scdney_mac_binary/scdney_0.1.2.tgz", repos = NULL)
```

# Reference