<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Sydney Precision Bioinformatics Group" />

<meta name="date" content="2019-03-05" />

<title>QC of single cell data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 64px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h2 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h3 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h4 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h5 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h6 {
  padding-top: 69px;
  margin-top: -69px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SingleCellPlus - HKU Workshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="qc.html">QC Process</a>
</li>
<li>
  <a href="scMerge.html">scMerge</a>
</li>
<li>
  <a href="downstream.html">Downstream Analysis</a>
</li>
<li>
  <a href="acknowledgement.html">Acknowledgement</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">QC of single cell data</h1>
<h4 class="author">Sydney Precision Bioinformatics Group</h4>
<h4 class="date">03/05/2019</h4>

</div>


<p>As discussed in 3rd May 2019, Yue Cao and Hani Kim will be jointly working on this part of the workshop.</p>
<p>However, the raw data will be prepared by Yingxin.</p>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>One of the first things you need to do when you have obtained single-cell RNA-seq data is to perform a quality check of the reads you have sequenced. This process is part of cleaning our expression matrix in preparation for data analysis. There are multiple quality control (QC) tools for sequencing data that can be used for both bulk and single-cell RNA-seq data. Here, we will work on performing various QCs and generate a report on the read quality of our data.</p>
</div>
<div id="load-in-packages" class="section level1">
<h1><span class="header-section-number">2</span> Load in packages</h1>
<pre class="r"><code>library(DropletUtils) #Droplet Utils is needed to plot the waterfall plot</code></pre>
</div>
<div id="the-liver-datasets" class="section level1">
<h1><span class="header-section-number">3</span> The Liver Datasets</h1>
<p>To illustrate cell QC, today we will be performing our analysis using single cell mouse liver dataset generated by Su et al. (2017). This liver dataset contains 507 cells at seven developmental stage between embryonic day 11.5 and postnatal day 2.5 The cells were sequenced using the Fluidigm C1 platform and the reads are paired end. The files are located in the ‘X’ folder in your working directory. These files are the raw count matrices.</p>
<p>Load the data and annotations: # Read in data</p>
<pre class="r"><code>data &lt;- read.csv(&quot;data/GSE87795_counts.csv&quot;)</code></pre>
<p>Inspect a small portion of the expression matrix.</p>
<pre class="r"><code>head(data[1:10])</code></pre>
<pre><code>##                    X E11.5_C01 E11.5_C03 E11.5_C05 E11.5_C06 E11.5_C07
## 1 ENSMUSG00000102693         0         0         0         0         0
## 2 ENSMUSG00000064842         0         0         0         0         0
## 3 ENSMUSG00000051951         0         0         0         0         0
## 4 ENSMUSG00000102851         0         0         0         0         0
## 5 ENSMUSG00000103377         0         0         0         0         0
## 6 ENSMUSG00000104017         0         0         0         0         0
##   E11.5_C09 E11.5_C11 E11.5_C12 E11.5_C14
## 1         0         0         0         0
## 2         0         0         0         0
## 3         0         0         0         0
## 4         0         0         0         0
## 5         0         0         0         0
## 6         0         0         0         0</code></pre>
<pre class="r"><code>dim(data)</code></pre>
<pre><code>## [1] 51918   508</code></pre>
<p>This dataset contains 51918 rows and 507 columns. Each row corresponds to a RNA molecule and each column correspond to an individual cell that was sequenced.</p>
<p>The first column records the name of the RNA molecule. We can assign it as the column name of the count matrix and then remove this column from the count matrix. Now the count matrix only contains numerical values and this makes subsequent analysis easier.</p>
<pre class="r"><code>names_data &lt;- data[,1]
rownames(data) &lt;- names_data
data &lt;- data[, -1]
head(data[1:10])</code></pre>
<pre><code>##                    E11.5_C01 E11.5_C03 E11.5_C05 E11.5_C06 E11.5_C07
## ENSMUSG00000102693         0         0         0         0         0
## ENSMUSG00000064842         0         0         0         0         0
## ENSMUSG00000051951         0         0         0         0         0
## ENSMUSG00000102851         0         0         0         0         0
## ENSMUSG00000103377         0         0         0         0         0
## ENSMUSG00000104017         0         0         0         0         0
##                    E11.5_C09 E11.5_C11 E11.5_C12 E11.5_C14 E11.5_C15
## ENSMUSG00000102693         0         0         0         0         0
## ENSMUSG00000064842         0         0         0         0         0
## ENSMUSG00000051951         0         0         0         0         0
## ENSMUSG00000102851         0         0         0         0         0
## ENSMUSG00000103377         0         0         0         0         0
## ENSMUSG00000104017         0         0         0         0         0</code></pre>
</div>
<div id="what-are-barcode" class="section level1">
<h1><span class="header-section-number">4</span> What are barcode</h1>
<p>Barcode are short DNA strands that are tagged on each cell. This allows identification of reads that comes from the same cell. In a count matrix barcode usually appear as column name. We can consider each barcode as each cell.</p>
</div>
<div id="remove-genes" class="section level1">
<h1><span class="header-section-number">5</span> Remove genes</h1>
<p>Some genes may not be expressed in any cell. We can remove these genes.</p>
</div>
<div id="calculate-quality-metrics" class="section level1">
<h1><span class="header-section-number">6</span> Calculate quality metrics</h1>
<p>UMI stands for unique molecular identifier. In the sequencing step, a UMI is attached to RNA molecules, such that RNA molecules with same UMI is considered as the coming from the same input molecule. This information is then needed when quantifying the number of RNA molecules detected per sample. The total number of RNA molecules per sample is called the library size of each sample.</p>
</div>
<div id="quality-control-of-single-cells" class="section level1">
<h1><span class="header-section-number">7</span> Quality Control of Single Cells</h1>
</div>
<div id="qc1-waterfall-plot" class="section level1">
<h1><span class="header-section-number">8</span> QC1: Waterfall Plot</h1>
<pre class="r"><code>br.out &lt;- barcodeRanks(data)
plot(br.out$rank, br.out$total, log=&quot;xy&quot;, xlab=&quot;Rank&quot;, ylab=&quot;Total&quot;, col=&quot;royalblue&quot;)
o &lt;- order(br.out$rank)
lines(br.out$rank[o], br.out$fitted[o], col=&quot;red&quot;)
abline(h=br.out$knee, col=&quot;dodgerblue&quot;, lty=2)
abline(h=br.out$inflection, col=&quot;forestgreen&quot;, lty=2)
legend(&quot;bottomleft&quot;, lty=2, col=c(&quot;dodgerblue&quot;, &quot;forestgreen&quot;),
        legend=c(&quot;knee&quot;, &quot;inflection&quot;))</code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-5-1.png" width="576" /></p>
</div>
<div id="qc2-library-size" class="section level1">
<h1><span class="header-section-number">9</span> QC2: Library size</h1>
<pre class="r"><code>options(scipen=999)
par(mfrow=c(1,2))
hist(colSums(data[, grepl(&quot;E11.5&quot;,colnames(data))])/1e3, breaks = 20, main = &quot;E11.5&quot;, ylab = &quot;number of cells&quot;, xlab = &quot;library size (thousands)&quot;, col=&quot;gray&quot;)
hist(colSums(data[, grepl(&quot;E13.5&quot;,colnames(data))])/1e3, breaks = 20, main = &quot;E13.5&quot;, ylab = &quot;number of cells&quot;, xlab = &quot;library size (thousands)&quot;, col=&quot;gray&quot;)</code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-6-1.png" width="576" /></p>
</div>
<div id="qc3-number-of-uniquely-expressed-genes" class="section level1">
<h1><span class="header-section-number">10</span> QC3: Number of uniquely expressed genes</h1>
<pre class="r"><code>par(mfrow=c(1,2))
hist(colSums(data[, grepl(&quot;E11.5&quot;,colnames(data))]!=0), breaks = 20, main = &quot;E11.5&quot;, ylab = &quot;number of cells&quot;, xlab = &quot;number of expressed genes&quot;, col=&quot;gray&quot;)
hist(colSums(data[, grepl(&quot;E13.5&quot;,colnames(data))]!=0), breaks = 20, main = &quot;E13.5&quot;, ylab = &quot;number of cells&quot;, xlab = &quot;number of expressed genes&quot;, col=&quot;gray&quot;)</code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-7-1.png" width="576" /></p>
</div>
<div id="qc5-contribution-by-top-50-expressed-genes" class="section level1">
<h1><span class="header-section-number">11</span> QC5: Contribution by top 50 expressed genes</h1>
</div>
<div id="qc6-mitochondrial-gene-expression" class="section level1">
<h1><span class="header-section-number">12</span> QC6: Mitochondrial gene expression</h1>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
