<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Sydney Precision Bioinformatics Group" />

<meta name="date" content="2019-06-18" />

<title>QC of single cell data</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 64px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h2 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h3 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h4 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h5 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h6 {
  padding-top: 69px;
  margin-top: -69px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SingleCellPlus - HKU Workshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="qc.html">QC Process</a>
</li>
<li>
  <a href="scMerge.html">scMerge</a>
</li>
<li>
  <a href="downstream.html">Downstream Analysis</a>
</li>
<li>
  <a href="acknowledgement.html">Acknowledgement</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">QC of single cell data</h1>
<h4 class="author">Sydney Precision Bioinformatics Group</h4>
<h4 class="date">18 June 2019</h4>

</div>


<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>One of the first steps in single-cell RNA-seq analysis is to perform a quality check of the data you have sequenced in an experiment. This process is a necessary step before we proceed to data merging and other downstream data analyses. There are multiple quality control (QC) tools for sequencing data that we can use for both bulk and single-cell RNA-seq data. Here, we will look at some QC measures and generate a report on the quality of our single-cell data.</p>
<div id="load-in-packages" class="section level2">
<h2><span class="header-section-number">1.1</span> Load in packages</h2>
<p>Load in packages required to generate the QC plots.</p>
<pre class="r"><code>library(DropletUtils)
library(dplyr)
library(ggplot2)
library(scales)
library(ggpubr)
library(stringr)
library(forcats)

theme_set(theme_classic(16))</code></pre>
</div>
<div id="the-mouse-liver-dataset-su-et-al.-2017" class="section level2">
<h2><span class="header-section-number">1.2</span> The mouse liver dataset (Su et al. 2017)</h2>
<p>For demonstration purpose, we will perform some QC checks on a single-cell mouse liver dataset generated by Su et al. (2017). This liver dataset contains 507 cells at seven developmental stages between embryonic day 11.5 and postnatal day 2.5. The cells were sequenced using the Fluidigm C1 platform and the reads are paired end. The file containing the raw count matrix is located in the <code>data</code> folder in the zip file you have downloaded.</p>
<p>The dataset of read counts is stored in a <code>csv</code> file (which is simply an Excel file). We can load the data into R using the <code>read.csv()</code> function. Here, we will use two options:</p>
<ul>
<li><code>header = TRUE</code> means that we will take the first row of the file as the column names.</li>
<li><code>row.names = 1</code> means that we will take the first column of the file as the row names.</li>
</ul>
<pre class="r"><code>liver = read.csv(file = &quot;data/GSE87795_counts.csv&quot;, 
                 header = TRUE,
                 row.names = 1)</code></pre>
<p>First, let’s check some basics of this data. This dataset contains 51918 rows and 507 columns. Each row corresponds to a gene, and each column corresponds to a cell. The first five rows and five columns of the dataset can be shown using <code>liver[1:5, 1:5]</code>.</p>
<pre class="r"><code>dim(liver)</code></pre>
<pre><code>## [1] 51918   507</code></pre>
<pre class="r"><code>liver[1:5, 1:5]</code></pre>
<pre><code>##                    E11.5_C01 E11.5_C03 E11.5_C05 E11.5_C06 E11.5_C07
## ENSMUSG00000102693         0         0         0         0         0
## ENSMUSG00000064842         0         0         0         0         0
## ENSMUSG00000051951         0         0         0         0         0
## ENSMUSG00000102851         0         0         0         0         0
## ENSMUSG00000103377         0         0         0         0         0</code></pre>
<p>Notice that the column names of this data contain information regarding the developmental stage of each cell when it was sequenced in the experiment. We can clean up this information using a short line of code.</p>
<p>With some code, we can produce a frequency table of the number of cells at each developmental stage.</p>
<pre class="r"><code>stage = str_split(colnames(liver), &quot;_&quot;) %&gt;% 
  sapply(&quot;[[&quot;, 1)

table(stage)</code></pre>
<pre><code>## stage
## E11.5 E12.5 E13.5 E14.5 E16.5 E18.5  P2.5 
##    70   100    71    99    78    58    31</code></pre>
</div>
<div id="quick-pre-processing-of-the-matrix" class="section level2">
<h2><span class="header-section-number">1.3</span> Quick Pre-processing of the Matrix</h2>
<p>A simple way to make single-cell datasets manageable for analysis is to remove genes that are not expressed in any cells (i.e. genes that have zero counts across all cells). Since these genes do not containn any measurable signals, removing them does not affect the downstream analysis. In addition, a smaller dataset actually improves the performance and speed of most algorithms.</p>
<pre class="r"><code>liver = liver[rowSums(liver) != 0, ]
dim(liver)</code></pre>
<pre><code>## [1] 37766   507</code></pre>
<p>Here, we first used <code>rowSums(liver)</code> to calculate the total counts for each gene across all cells. <code>rowSums(liver) != 0</code> then looks at if the total count is zero or not. And we will subset only those rows in the data (remember, rows of the data are genes) that are not purely zeroes.</p>
<h4>
<span class="glyphicon glyphicon-education" aria-hidden="true"></span> Quick quiz
</h4>
<p>How many genes do we have now?</p>
</div>
<div id="quality-control-of-single-cell-data" class="section level2">
<h2><span class="header-section-number">1.4</span> Quality Control of Single Cell Data</h2>
<div id="qc1-waterfall-plot" class="section level3">
<h3><span class="header-section-number">1.4.1</span> QC1: Waterfall Plot</h3>
<p>The waterfall plot shows the log-count against the log-rank of each barcode. The barcodes are ranked based on the number of count each barcode has. The plots show two points of interest:</p>
<ul>
<li><strong>inflection point</strong> is the point on the curve where the first derivative is minimised</li>
<li><strong>knee point</strong> is the point where the second derivative is minimised</li>
</ul>
<p>These two points potentially indicate empty droplets (background barcodes). If the number of count in each cell drops rapidly for those cells beyond knee point, then it may suggest empty reads.</p>
<pre class="r"><code>barcodeRank = DropletUtils::barcodeRanks(liver)
barcodeRank.data = as.data.frame(barcodeRank)
barcodeRank.points = data.frame(
  type = c(&quot;inflection&quot;, &quot;knee&quot;),
  value = c(barcodeRank@metadata$inflection, barcodeRank@metadata$knee))


ggplot(data = barcodeRank.data, aes(x = rank, y = total)) +
  geom_point() +
  geom_hline(data = barcodeRank.points, 
             aes(yintercept = value, 
                 colour = type), linetype = 2) +
  scale_y_log10(breaks = trans_breaks(&quot;log10&quot;, function(x) 10^x),
                labels = trans_format(&quot;log10&quot;, math_format(10^.x))) +
  labs(title = &quot;Waterfall plot of read counts (log)&quot;, 
       x = &quot;Rank&quot;, 
       y = &quot;Log counts&quot;)</code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-5-1.png" width="576" /></p>
<h4>
<span class="glyphicon glyphicon-education" aria-hidden="true"></span> Quick quiz
</h4>
<p>Based on the waterfall plot, should we filter out any cells?</p>
</div>
<div id="qc2-library-size" class="section level3">
<h3><span class="header-section-number">1.4.2</span> QC2: Library Size</h3>
<p>Next, we can evaluate the total library size, that is the collective read counts per cell. The library size may be used as an indicator of sequencing depth.</p>
<pre class="r"><code>library.plotdf.full = data.frame(
  cellName = colnames(liver),
  stage = stage, 
  library = colSums(liver),
  nGenes = colSums(liver != 0)
) 

library.plotdf = library.plotdf.full %&gt;% 
 dplyr::filter(stage %in% c(&quot;E11.5&quot;, &quot;E12.5&quot;, &quot;E13.5&quot;))


library.stage.mean = library.plotdf %&gt;% 
  group_by(stage) %&gt;% 
  dplyr::summarise(stage.library.mean = mean(library), 
                   stage.nGenes.exprs = mean(nGenes))


library.stage.mean</code></pre>
<pre><code>## # A tibble: 3 x 3
##   stage stage.library.mean stage.nGenes.exprs
##   &lt;fct&gt;              &lt;dbl&gt;              &lt;dbl&gt;
## 1 E11.5           2838642.              9262.
## 2 E12.5           3094663.              9346.
## 3 E13.5           3639297.              9813.</code></pre>
<pre class="r"><code>ggplot(library.plotdf, aes(x = library, colour = stage, fill = stage)) +
  geom_histogram(aes(y = ..density..), colour = &quot;black&quot;, fill = &quot;white&quot;, bins = 30) + 
  geom_density(alpha = 0.4) + 
  geom_vline(data = library.stage.mean, 
             aes(xintercept = stage.library.mean), 
             colour = &quot;blue&quot;, linetype = &quot;dashed&quot;, size = 1.5) +
  labs(title = &quot;Histogram of library size&quot;, 
       x = &quot;library size&quot;, 
       y = &quot;density&quot;) + 
  facet_wrap(~ stage, ncol = 3) </code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-6-1.png" width="768" /></p>
<p>When we compare the distribution of librarzy size for all cells at three developmental stages, we can see that mean library size is lowest in E11.5 cells (mean=2838642) and highest in E13.5 cells (mean=3639297). E12.5 cells (mean=3094663) have a greater distribution of cells with few reads. Single cells with very few reads are likely to have failed to capture the transcriptome of a cell, and thus should be filtered.</p>
</div>
</div>
<div id="qc3-number-of-uniquely-expressed-genes" class="section level2">
<h2><span class="header-section-number">1.5</span> QC3: Number of Uniquely Expressed Genes</h2>
<p>In addition to sequencing depth, we can also filter for cells that express genes with sufficiently good coverage of the transcriptome. This is to ensure that we capture cells with not only sufficient reads but also those that have reads that are relatively evenly distributed across the transcriptome. To this end, we measure the number of uniquely expressed genes in individual cells.</p>
<pre class="r"><code>ggplot(library.plotdf, 
       aes(x = nGenes, colour = stage, fill = stage)) +
  geom_histogram(aes(y = ..density..), colour = &quot;black&quot;, fill = &quot;white&quot;, bins = 30) + 
  geom_density(alpha = 0.4) + 
  geom_vline(data = library.stage.mean,
             aes(xintercept = stage.nGenes.exprs), 
             colour = &quot;blue&quot;, linetype = &quot;dashed&quot;, size = 1.5) +
  labs(title = &quot;Histogram of number of unique genes expressed&quot;, 
       x = &quot;Number of unique genes expressed&quot;, 
       y = &quot;density&quot;) + 
  facet_wrap(~ stage, ncol = 3) </code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-7-1.png" width="576" /></p>
<p>From above, we can see that most cells have between 9000-10,000 detected genes. This is roughly the amount expected for high-depth scRNA-seq; however, the expected value may vary between experimental protocols, sequencing depth, and cell type used. Note that, unlike library size, a “heavy tail” can be observed for each of the distribution, indicating an unequal detection of genes across populations. Filtering out cells with low number of genes identified is another way to improve the quality of the dataset.</p>
<div id="qc4-mitochondrial-gene-expression" class="section level3">
<h3><span class="header-section-number">1.5.1</span> QC4: Mitochondrial Gene Expression</h3>
<p>Unhealthy or dying cells are associated with high expression of mitochondrial genes, and thus the proportion of mitochondrial expression is commonly used as a quality metric.</p>
<pre class="r"><code>mito.genes &lt;- c(&quot;ENSMUSG00000064336&quot;, &quot;ENSMUSG00000064337&quot;, &quot;ENSMUSG00000064338&quot;, &quot;ENSMUSG00000064339&quot;, &quot;ENSMUSG00000064340&quot;, &quot;ENSMUSG00000064341&quot;, &quot;ENSMUSG00000064342&quot;, &quot;ENSMUSG00000064343&quot;, &quot;ENSMUSG00000064344&quot;, &quot;ENSMUSG00000064345&quot;, &quot;ENSMUSG00000064346&quot;, &quot;ENSMUSG00000064347&quot;, &quot;ENSMUSG00000064348&quot;, &quot;ENSMUSG00000064349&quot;, &quot;ENSMUSG00000064350&quot;, &quot;ENSMUSG00000064351&quot;, &quot;ENSMUSG00000064352&quot;, &quot;ENSMUSG00000064353&quot;, &quot;ENSMUSG00000064354&quot;, &quot;ENSMUSG00000064355&quot;, &quot;ENSMUSG00000064356&quot;, &quot;ENSMUSG00000064357&quot;, &quot;ENSMUSG00000064358&quot;, &quot;ENSMUSG00000064359&quot;, &quot;ENSMUSG00000064360&quot;, &quot;ENSMUSG00000064361&quot;, &quot;ENSMUSG00000065947&quot;, &quot;ENSMUSG00000064363&quot;, &quot;ENSMUSG00000064364&quot;, &quot;ENSMUSG00000064365&quot;, &quot;ENSMUSG00000064366&quot;, &quot;ENSMUSG00000064367&quot;, &quot;ENSMUSG00000064368&quot;, &quot;ENSMUSG00000064369&quot;, &quot;ENSMUSG00000064370&quot;, &quot;ENSMUSG00000064371&quot;, &quot;ENSMUSG00000064372&quot;, &quot;ENSMUSG00000096105&quot;)


liver.mito = liver[mito.genes, ]

mito.plotdf =  data.frame(
  cellName = colnames(liver.mito),
  stage = stage, 
  mito.percent = colSums(liver.mito)/colSums(liver) * 100
) %&gt;% dplyr::filter(stage %in% c(&quot;E11.5&quot;, &quot;E12.5&quot;, &quot;E13.5&quot;))

mito.mean = mito.plotdf %&gt;% 
  group_by(stage) %&gt;% 
  dplyr::summarise(mean.mito.percent = mean(mito.percent))

mito.mean</code></pre>
<pre><code>## # A tibble: 3 x 2
##   stage mean.mito.percent
##   &lt;fct&gt;             &lt;dbl&gt;
## 1 E11.5              7.61
## 2 E12.5              6.99
## 3 E13.5              5.35</code></pre>
<pre class="r"><code>g1 = ggplot(mito.plotdf, 
       aes(x = mito.percent, colour = stage, fill = stage)) +
  geom_histogram(aes(y = ..density..), colour = &quot;black&quot;, fill = &quot;white&quot;, bins = 30) + 
  geom_density(alpha = 0.4) + 
  geom_vline(data = mito.mean,
             aes(xintercept = mean.mito.percent), 
             colour = &quot;blue&quot;, linetype = &quot;dashed&quot;, size = 1.5) +
  labs(x = &quot;mitochondrial percentage&quot;, 
       y = &quot;density&quot;) + 
  scale_x_continuous(limits = c(0, 15)) +
  facet_wrap(~ stage, nrow = 3) 


g2 = ggplot(mito.plotdf, 
            aes(x = stage, 
                y = mito.percent, 
                fill = stage)) +
  geom_violin(alpha = 0.8, draw_quantiles = 0.5) + 
  labs(y = &quot;mitochondrial percentage&quot;)

ggpubr::ggarrange(g1, g2, ncol=  2, common.legend = TRUE, legend = &quot;right&quot;)</code></pre>
<pre><code>## Warning: Removed 11 rows containing non-finite values (stat_bin).</code></pre>
<pre><code>## Warning: Removed 11 rows containing non-finite values (stat_density).</code></pre>
<pre><code>## Warning: Removed 6 rows containing missing values (geom_bar).</code></pre>
<pre><code>## Warning: Removed 11 rows containing non-finite values (stat_bin).</code></pre>
<pre><code>## Warning: Removed 11 rows containing non-finite values (stat_density).</code></pre>
<pre><code>## Warning: Removed 6 rows containing missing values (geom_bar).</code></pre>
<pre><code>## Warning in regularize.values(x, y, ties, missing(ties)): collapsing to
## unique &#39;x&#39; values</code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-8-1.png" width="960" /></p>
<p>Filtering cells with greater than 10% of mitochondrial gene expression would remove cells that have abnormally high mitochondrial gene expression.</p>
<h4>
<span class="glyphicon glyphicon-education" aria-hidden="true"></span> Quick quiz
</h4>
<ol style="list-style-type: decimal">
<li>Based on the figure above, would you say most cells are unhealthy or healthy?<br />
</li>
<li>How does the percentage of mitochondrial gene expression change over time point ?</li>
</ol>
</div>
</div>
<div id="qc5-contribution-by-top-50-expressed-genes" class="section level2">
<h2><span class="header-section-number">1.6</span> QC5: Contribution by Top 50 Expressed Genes</h2>
<p>As well as the above metrics, it is often instructive to know the proportions of the reads consumed by the top 50 expressed genes. Moreover, an assessment of the top gene list before and after filtering can give you a good indication of whether the filtering has been successful, as the top genes are often occupied by mitochondrial genes from unhealthy cells.</p>
<pre class="r"><code>liver.percent = sweep(liver, 2, STATS = colSums(liver), FUN = &quot;/&quot;) * 100

liver.percent %&gt;% colSums %&gt;% head</code></pre>
<pre><code>## E11.5_C01 E11.5_C03 E11.5_C05 E11.5_C06 E11.5_C07 E11.5_C09 
##       100       100       100       100       100       100</code></pre>
<pre class="r"><code>liver.percent.split = split.data.frame(x = t(liver.percent), f = stage) %&gt;% lapply(t)

liver.percent.split.rowMeans = liver.percent.split %&gt;% lapply(rowMeans)


e115.top50 = liver.percent.split.rowMeans$E11.5 %&gt;% sort(decreasing = TRUE) %&gt;% head(50) %&gt;% names
e115.top50</code></pre>
<pre><code>##  [1] &quot;ENSMUSG00000101111&quot; &quot;ENSMUSG00000064358&quot; &quot;ENSMUSG00000064351&quot;
##  [4] &quot;ENSMUSG00000019505&quot; &quot;ENSMUSG00000029580&quot; &quot;ENSMUSG00000064370&quot;
##  [7] &quot;ENSMUSG00000064354&quot; &quot;ENSMUSG00000102070&quot; &quot;ENSMUSG00000037742&quot;
## [10] &quot;ENSMUSG00000064339&quot; &quot;ENSMUSG00000101249&quot; &quot;ENSMUSG00000062825&quot;
## [13] &quot;ENSMUSG00000100862&quot; &quot;ENSMUSG00000106106&quot; &quot;ENSMUSG00000063632&quot;
## [16] &quot;ENSMUSG00000066551&quot; &quot;ENSMUSG00000008348&quot; &quot;ENSMUSG00000068823&quot;
## [19] &quot;ENSMUSG00000064341&quot; &quot;ENSMUSG00000064357&quot; &quot;ENSMUSG00000057113&quot;
## [22] &quot;ENSMUSG00000023944&quot; &quot;ENSMUSG00000044285&quot; &quot;ENSMUSG00000069662&quot;
## [25] &quot;ENSMUSG00000037894&quot; &quot;ERCC-00002&quot;         &quot;ERCC-00074&quot;        
## [28] &quot;ENSMUSG00000054766&quot; &quot;ENSMUSG00000016559&quot; &quot;ENSMUSG00000001525&quot;
## [31] &quot;ERCC-00096&quot;         &quot;ENSMUSG00000071072&quot; &quot;ENSMUSG00000064363&quot;
## [34] &quot;ENSMUSG00000032518&quot; &quot;ENSMUSG00000060036&quot; &quot;ENSMUSG00000057666&quot;
## [37] &quot;ENSMUSG00000015656&quot; &quot;ENSMUSG00000022816&quot; &quot;ENSMUSG00000067274&quot;
## [40] &quot;ENSMUSG00000021427&quot; &quot;ENSMUSG00000072235&quot; &quot;ERCC-00130&quot;        
## [43] &quot;ENSMUSG00000071866&quot; &quot;ENSMUSG00000022048&quot; &quot;ENSMUSG00000021546&quot;
## [46] &quot;ENSMUSG00000046434&quot; &quot;ENSMUSG00000032399&quot; &quot;ENSMUSG00000023004&quot;
## [49] &quot;ENSMUSG00000058558&quot; &quot;ENSMUSG00000029836&quot;</code></pre>
<pre class="r"><code>e115.liver.percent = liver.percent[e115.top50, stage == &quot;E11.5&quot;] %&gt;% 
  as.matrix %&gt;% 
  reshape2::melt(varnames = c(&quot;gene.name&quot;, &quot;cell.name&quot;)) %&gt;% 
  dplyr::mutate(gene.name = fct_reorder(gene.name, value, mean))

e115.liver.percent %&gt;% 
  ggplot(aes(x = gene.name, y = value, colour = gene.name)) +
  geom_violin() +
  geom_jitter(alpha = 0.2) +
  coord_flip() +
  labs(x = &quot;percent of library&quot;, 
       y = &quot;genes&quot;) +
  theme(legend.position = &quot;none&quot;)</code></pre>
<p><img src="qc_files/figure-html/unnamed-chunk-9-1.png" width="768" /></p>
<!-- ```{r} -->
<!-- liver.percent.melt = liver.percent %>% -->
<!--   as.matrix %>%  -->
<!--   reshape2::melt(varnames = c("gene.name", "cell.name"), value.name = "percent") %>%  -->
<!--   dplyr::mutate(stage = str_split(cell.name, "_") %>% sapply("[[", 1))  -->
<!-- e115.percent = liver.percent.melt %>%  -->
<!--   dplyr::filter(stage == "E11.5") %>%  -->
<!--   dplyr::mutate(gene.name = forcats::fct_reorder(gene.name, percent, mean, .desc = TRUE)) %>%  -->
<!--   dplyr::filter(as.integer(gene.name) <= 50) -->
<!-- e115.percent %>%  -->
<!--   ggplot(aes(x = fct_rev(gene.name), y = percent, colour = gene.name)) + -->
<!--   geom_violin() + -->
<!--   geom_jitter(alpha = 0.2) + -->
<!--   coord_flip() + -->
<!--   labs(x = "percent",  -->
<!--        y = "") -->
<!--   theme(legend.position = "none") -->
<!-- ``` -->
<p>The plot above show that there are four spike-in genes within the top 50 genes. The presence of multiple spike-ins at the top of the list suggests that the concentration of spike-ins added to the experiment may need to be optimized for subsequent experiments. Overall, the relatively flat distributions indicate good coverage of the transcriptome.</p>
<div id="filtering" class="section level3">
<h3><span class="header-section-number">1.6.1</span> Filtering</h3>
<p>Here we will give a demonstration of how you can remove some of the cells from the dataset.<br />
Recall in the waterfall plot, we have identified 2 outlier cells with low read counts. Here, we will illustrate how you can remove these two cells from the dataset.</p>
<p>In the code provided below, originally there was 507 cells in the dataset, the 2 outlier cells would have rank 506 and 507 (largest rank means least number of reads). We look at which cells has rank &gt;= 506 and use the row index to subset the dataset. The which() function gives you the row index and the minus sign ‘-’ tells R to remove the given row index from the dataset.</p>
<p>Finally we should check that the dimension of the dataset is correct after filtering.</p>
<pre class="r"><code>liver_new &lt;- liver[, barcodeRank.data$rank &lt;= 506]
dim(liver_new)</code></pre>
<pre><code>## [1] 37766   506</code></pre>
<p>[Extension]. Try removing the cells with abnormally high mitochondrial gene expression</p>
</div>
</div>
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">2</span> Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] forcats_0.4.0               stringr_1.4.0              
##  [3] scales_1.0.0                DropletUtils_1.4.1         
##  [5] plyr_1.8.4                  ggpubr_0.2                 
##  [7] magrittr_1.5                viridis_0.5.1              
##  [9] viridisLite_0.3.0           MAST_1.10.0                
## [11] ggplot2_3.1.1               cluster_2.0.9              
## [13] Rtsne_0.15                  mclust_5.4.3               
## [15] scdney_0.1.2                edgeR_3.26.4               
## [17] limma_3.40.2                dplyr_0.8.1                
## [19] SingleCellExperiment_1.6.0  SummarizedExperiment_1.14.0
## [21] DelayedArray_0.10.0         BiocParallel_1.18.0        
## [23] matrixStats_0.54.0          Biobase_2.44.0             
## [25] GenomicRanges_1.36.0        GenomeInfoDb_1.20.0        
## [27] IRanges_2.18.1              S4Vectors_0.22.0           
## [29] BiocGenerics_0.30.0        
## 
## loaded via a namespace (and not attached):
##   [1] snow_0.4-3             backports_1.1.4        Hmisc_4.2-0           
##   [4] blme_1.0-4             igraph_1.2.4.1         lazyeval_0.2.2        
##   [7] splines_3.6.0          amap_0.8-17            digest_0.6.19         
##  [10] foreach_1.4.4          htmltools_0.3.6        fansi_0.4.0           
##  [13] checkmate_1.9.3        doParallel_1.0.14      recipes_0.1.5         
##  [16] gower_0.2.1            R.utils_2.8.0          colorspace_1.4-1      
##  [19] pan_1.6                xfun_0.7               crayon_1.3.4          
##  [22] RCurl_1.95-4.12        lme4_1.1-21            zeallot_0.1.0         
##  [25] survival_2.44-1.1      iterators_1.0.10       glue_1.3.1            
##  [28] gtable_0.3.0           ipred_0.9-9            zlibbioc_1.30.0       
##  [31] XVector_0.24.0         Rhdf5lib_1.6.0         HDF5Array_1.12.1      
##  [34] jomo_2.6-8             abind_1.4-5            mvtnorm_1.0-10        
##  [37] Rcpp_1.0.1             htmlTable_1.13.1       dqrng_0.2.1           
##  [40] foreign_0.8-71         Formula_1.2-3          lava_1.6.5            
##  [43] prodlim_2018.04.18     htmlwidgets_1.3        RColorBrewer_1.1-2    
##  [46] ellipsis_0.1.0         acepack_1.4.1          mice_3.5.0            
##  [49] pkgconfig_2.0.2        R.methodsS3_1.7.1      manipulate_1.0.1      
##  [52] nnet_7.3-12            locfit_1.5-9.1         utf8_1.1.4            
##  [55] caret_6.0-84           tidyselect_0.2.5       labeling_0.3          
##  [58] rlang_0.3.4            reshape2_1.4.3         munsell_0.5.0         
##  [61] tools_3.6.0            cli_1.1.0              generics_0.0.2        
##  [64] broom_0.5.2            ggridges_0.5.1         evaluate_0.14         
##  [67] yaml_2.2.0             ModelMetrics_1.2.2     knitr_1.23            
##  [70] purrr_0.3.2            randomForest_4.6-14    mitml_0.3-7           
##  [73] dendextend_1.12.0      packrat_0.5.0          nlme_3.1-140          
##  [76] R.oo_1.22.0            compiler_3.6.0         rstudioapi_0.10       
##  [79] e1071_1.7-2            tibble_2.1.3           DescTools_0.99.28     
##  [82] stringi_1.4.3          lattice_0.20-38        Matrix_1.2-17         
##  [85] nloptr_1.2.1           vctrs_0.1.0            pillar_1.4.1          
##  [88] BiocManager_1.30.4     data.table_1.12.2      cowplot_0.9.4         
##  [91] bitops_1.0-6           R6_2.4.0               latticeExtra_0.6-28   
##  [94] gridExtra_2.3          codetools_0.2-16       boot_1.3-22           
##  [97] MASS_7.3-51.4          assertthat_0.2.1       rhdf5_2.28.0          
## [100] withr_2.1.2            GenomeInfoDbData_1.2.1 expm_0.999-4          
## [103] doSNOW_1.0.16          grid_3.6.0             rpart_4.1-15          
## [106] timeDate_3043.102      tidyr_0.8.3            class_7.3-15          
## [109] minqa_1.2.4            rmarkdown_1.13         clusteval_0.1         
## [112] lubridate_1.7.4        base64enc_0.1-3</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
