---
title: "QC of single cell data"
author: "Sydney Precision Bioinformatics Group"
date: "03/05/2019"
output:
  html_document:
    code_folding: hide
    number_sections: yes
    toc: yes
    fig_height: 8
    fig_width: 8
    toc_float:
      collapsed: true
      smooth_scroll: false
---


As discussed in 3rd May 2019, Yue Cao and Hani Kim will be jointly working on this part of the workshop. 

However, the raw data will be prepared by Yingxin. 


# Introduction
One of the first things you need to do when you have obtained single-cell RNA-seq data is to perform a quality check of the reads you have sequenced. This process is part of cleaning our expression matrix in preparation for data analysis. There are multiple quality control (QC) tools for sequencing data that can be used for both bulk and single-cell RNA-seq data. Here, we will work on performing various QCs and generate a report on the read quality of our data. 




# Load in packages 
```{r warning=FALSE, message=FALSE}
library(DropletUtils) #Droplet Utils is needed to plot the waterfall plot
```

# The Liver Datasets
To illustrate cell QC, today we will be performing our analysis using single cell mouse liver dataset generated by Su et al. (2017). This liver dataset contains 507 cells at seven developmental stage between embryonic day 11.5 and postnatal day 2.5  The cells were sequenced using the Fluidigm C1 platform and the reads are paired end. The files are located in the ‘X’ folder in your working directory. These files are the raw count matrices. 

Load the data and annotations:
# Read in data

```{r}

data <- read.csv("data/GSE87795_counts.csv")

```


Inspect a small portion of the expression matrix.

```{r}
head(data[1:10])
dim(data)
```
This dataset contains 51918 rows and 507 columns. Each row corresponds to a RNA molecule and each column correspond to an individual cell that was sequenced. 


The first column records the name of the RNA molecule. We can assign it as the column name of the count matrix and then remove this column from the count matrix. Now the count matrix only contains numerical values and this makes subsequent analysis easier. 


```{r}
names_data <- data[,1]
rownames(data) <- names_data
data <- data[, -1]
head(data[1:10])
```


# What are barcode
Barcode are short DNA strands that are tagged on each cell. This allows identification of reads that comes from the same cell. In a count matrix barcode usually appear as column name. We can consider each barcode as each cell.  

# Remove genes
Some genes may not be expressed in any cell. We can remove these genes. 


# Calculate quality metrics
UMI stands for unique molecular identifier. In the sequencing step, a UMI is attached to RNA molecules, such that RNA molecules with same UMI is considered as the coming from the same input molecule. This information is then needed when quantifying the number of RNA molecules detected per sample. The total number of RNA molecules per sample is called the library size of each sample. 


# Quality Control of Single Cells

# QC1: Waterfall Plot

```{r fig.height=6, fig.width=6}

br.out <- barcodeRanks(data)
plot(br.out$rank, br.out$total, log="xy", xlab="Rank", ylab="Total", col="royalblue")
o <- order(br.out$rank)
lines(br.out$rank[o], br.out$fitted[o], col="red")
abline(h=br.out$knee, col="dodgerblue", lty=2)
abline(h=br.out$inflection, col="forestgreen", lty=2)
legend("bottomleft", lty=2, col=c("dodgerblue", "forestgreen"),
        legend=c("knee", "inflection"))
```


# QC2: Library size

```{r fig.height=4, fig.width=6}
options(scipen=999)
par(mfrow=c(1,2))
hist(colSums(data[, grepl("E11.5",colnames(data))])/1e3, breaks = 20, main = "E11.5", ylab = "number of cells", xlab = "library size (thousands)", col="gray")
hist(colSums(data[, grepl("E13.5",colnames(data))])/1e3, breaks = 20, main = "E13.5", ylab = "number of cells", xlab = "library size (thousands)", col="gray")

```

# QC3: Number of uniquely expressed genes


```{r fig.height=4, fig.width=6}

par(mfrow=c(1,2))
hist(colSums(data[, grepl("E11.5",colnames(data))]!=0), breaks = 20, main = "E11.5", ylab = "number of cells", xlab = "number of expressed genes", col="gray")
hist(colSums(data[, grepl("E13.5",colnames(data))]!=0), breaks = 20, main = "E13.5", ylab = "number of cells", xlab = "number of expressed genes", col="gray")

```




# QC5: Contribution by top 50 expressed genes
```{r}

```


# QC6: Mitochondrial gene expression
```{r}

```

