---
title: "scMerge package"
author: "Sydney Precision Bioinformatics Group"
date: "03/05/2019"
output: html_document
---

```{r, echo = FALSE}
t1 = Sys.time()
```


As discussed in 3rd May 2019, Kevin will work on this part of the workshop.

However, the raw data will be prepared by Yingxin and QC of the data are yet to be done. 

# Introduction 

scMerge is a method developed by the Sydney Precision Bioinformatics Group. It aims to merge multiple scRNA-Seq data so that researchers can look for biological signals on data pooled from multiple sources. The key to achieving a good merge is to remove the **noise**. 

Here are some plots where the raw data is separated by the data-batches but not cell types. 


scMerge corrects data-batch separations by 

1. identifying cells of identical cell types and pooling these together
2. identitying data-batch noises using Stably Expressed Genes (SEGs) and removing these from the data. 

Thus, scMerge can be thought of a two-step process!

# Supervised scMerge

We will begin with a liver dataset. The data are stored as four separate RDS files (this is similar to a `RData` file, but only one data can be stored in each RDS file). 




| Name  | ID       | Author | DOI or URL                | Protocol   | Organism | Tissue | # of cell types | # of cells | # of batches |
|-------|----------|--------|---------------------------|------------|----------|--------|-----------------|------------|--------------|
| Liver | GSE87795 | Su     | 10.1186/s12864-017-4342-x | SMARTer/C1 | Mouse    | Liver  | 6               | 1236       |              |
|       | GSE90047 | Yang   | 10.1002/hep.29353         | Smart-Seq2 |          |        | 2               |            |              |
|       | GSE87038 | Dong   | 10.1186/s13059-018-1416-2 | STRT-seq   |          |        | 3               |            |              |
|       | GSE96981 | Camp   | 10.1038/nature22796       | SMARTer/C1 |          |        | 3               |            |              |


```{r}
su = readRDS("data/sce_GSE87795.rds")
yang = readRDS("data/sce_GSE90047.rds")
dong = readRDS("data/sce_GSE87038.rds")
camp = readRDS("data/sce_GSE96981.rds")
sce_list = list(su = su, yang = yang, dong = dong, camp = camp)

```


```{r}
library(scMerge)
library(scater)
library(dplyr)


purrr::map(sce_list, ~ colData(.x) %>% colnames)

sce_combine = scMerge::sce_cbind(sce_list = sce_list, method = "union", colData_names = c("cellTypes", "stage"))
sce_combine
```



```{r}
data("segList_ensemblGeneID", package = "scMerge")

sce_combine <- scMerge(sce_combine = sce_combine, 
                       ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
                       kmeansK = c(6,2,3,3),
                       exprs = "logcounts",
                       hvg_exprs = "counts",
                       marker = NULL,
                       ruvK = 20,
                       cell_type = NULL,
                       dist = "cor",
                       replicate_prop = 1,
                       WV = sce_combine$stage,
                       # WV_marker = c("Afp","Alb","Epcam"),
                       WV_marker = c("ENSMUSG00000045394","ENSMUSG00000054932","ENSMUSG00000045394"),
                       return_all_RUV = FALSE,
                       assay_name = "scMerge_scSEG",
                       verbose = TRUE,
                       fast_svd = FALSE)
```



```{r}
scater::plotTSNE(sce_combine, 
                colour_by = 'cellTypes', shape = 'batch',
                run_args = list(exprs_values = 'logcounts'))

scater::plotTSNE(sce_combine, 
                colour_by = 'cellTypes', shape = 'batch',
                run_args = list(exprs_values = 'scMerge_scSEG'))
```


# Unsupervised scMerge

# Extension: semi-supervised scMerge


```{r, echo = FALSE}
t2 = Sys.time()
```



# Session Info
```{r}
sessionInfo()
t2 - t1
```

