---
title: "scMerge package"
author: "Sydney Precision Bioinformatics Group"
date: "03/05/2019"
output:
  html_document:
    code_folding: hide
    fig_height: 12
    fig_width: 12
    toc: yes
    number_sections: true
    toc_depth: 3
    toc_float: yes
---

# Jean's comments

1. `str` and `fct` manipulations as blackbox. 
1. Change "batch" into "data sources"/"datasets"
1. Separate plotly plots into two
1. Only keep Su and Yang. 
1. Dropping E10.5 and E16.5


```{r, echo = FALSE}
t1 = Sys.time()
```

# Introduction 

scMerge is a method developed by the Sydney Precision Bioinformatics Group. It aims to merge multiple scRNA-Seq data so that researchers can look for biological signals on data pooled from multiple sources. The key to achieving a good merge is to remove the **noise**. 

Here are some plots where the raw data is separated by the data-batches but not cell types. 


scMerge corrects data-batch separations by 

1. identifying cells of identical cell types and pooling these together
2. identitying data-batch noises using Stably Expressed Genes (SEGs) and removing these from the data. 

Thus, scMerge can be thought of a two-step process!

# Supervised scMerge

We will begin with a liver dataset. The data are stored as four separate RDS files (this is similar to a `RData` file, but only one data can be stored in each RDS file). 




| Name  | ID       | Author | DOI or URL                | Protocol   | Organism | Tissue | # of cell types | # of cells | # of batches |
|-------|----------|--------|---------------------------|------------|----------|--------|-----------------|------------|--------------|
| Liver | GSE87795 | Su     | 10.1186/s12864-017-4342-x | SMARTer/C1 | Mouse    | Liver  | 6               | 1236       | 6            |
|       | GSE90047 | Yang   | 10.1002/hep.29353         | Smart-Seq2 |          |        | 2               |            | 2            |
|       | GSE87038 | Dong   | 10.1186/s13059-018-1416-2 | STRT-seq   |          |        | 4               |            | 3            |
|       | GSE96981 | Camp   | 10.1038/nature22796       | SMARTer/C1 |          |        | 3               |            | 3            |


```{r}
su = readRDS("data/sce_GSE87795.rds")
yang = readRDS("data/sce_GSE90047.rds")
# dong = readRDS("data/sce_GSE87038.rds")
# camp = readRDS("data/sce_GSE96981.rds")
sce_list = list(
  su = su, 
  yang = yang
  # dong = dong, 
  # camp = camp
)
```


```{r, message=FALSE}
library(scMerge)
library(scater)
library(dplyr)
library(gridExtra)
library(plotly)
library(stringr)
library(forcats)

library(tibble)
library(dplyr)


# purrr::map(sce_list, ~ colData(.x) %>% colnames)

sce_combine = scMerge::sce_cbind(sce_list = sce_list, 
                                 method = "union", 
                                 colData_names = c("cellTypes", "stage"))


cat("Please copy and paste this")

sce_combine = sce_combine[,stringr::str_detect(sce_combine$cellTypes, "cholangiocyte|Epithelial Cell|hepatoblast|Hepatoblast|Mesenchymal Cell|Endothelial Cell|Hematopoietic|Immune cell|Stellate Cell")]

colData(sce_combine)$cellTypes = forcats::fct_recode(
  colData(sce_combine)$cellTypes, 
  Hepatoblast = "hepatoblast/hepatocyte"
  # Hepatoblast = "Hepatoblasts"
) %>% 
  droplevels()

sce_combine = sce_combine[rowSums(SingleCellExperiment::counts(sce_combine)) != 0,
                          colSums(SingleCellExperiment::counts(sce_combine)) != 0]
```



```{r, fig.width=12, fig.height=6, message=FALSE}
table(sce_combine$cellTypes, sce_combine$batch)

table(sce_combine$cellTypes, sce_combine$stage)

table(sce_combine$batch, sce_combine$stage)

pca_logcounts_cellTypes = scater::plotPCA(sce_combine, 
                                          colour_by = "cellTypes",
                                          run_args = list(exprs_values = "logcounts"))

pca_logcounts_batch = scater::plotPCA(sce_combine, 
                                      colour_by = "batch",
                                      run_args = list(exprs_values = "logcounts"))

grid.arrange(pca_logcounts_cellTypes, pca_logcounts_batch, nrow = 1)
```



```{r, fig.width=6, fig.height=6}
data("segList_ensemblGeneID", package = "scMerge")

sce_scMerge <- scMerge(
  sce_combine = sce_combine,
  ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
  cell_type = sce_combine$cellTypes,
  replicate_prop = 1,
  assay_name = "scMerge_supervised",
  verbose = TRUE)
```


```{r, fig.width=12, fig.height=6, message=FALSE}
sce_scMerge

pca_scMerge_supervised_cellTypes = scater::plotPCA(sce_scMerge, 
                                                   colour_by = "cellTypes",
                                                   run_args = list(exprs_values = "scMerge_supervised"))

pca_scMerge_supervised_batch = scater::plotPCA(sce_scMerge, 
                                               colour_by = "batch",
                                               run_args = list(exprs_values = "scMerge_supervised"))

grid.arrange(pca_scMerge_supervised_cellTypes, pca_scMerge_supervised_batch, nrow = 1)
```

# Unsupervised scMerge

```{r, fig.width=6, fig.height=6}
sce_scMerge <- scMerge(
  sce_combine = sce_scMerge,
  ctl = which(rownames(sce_scMerge) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
  kmeansK = c(6,2),
  replicate_prop = 1,
  assay_name = "scMerge_unsupervised",
  verbose = TRUE)
```


```{r, fig.width=12, fig.height=6, message=FALSE}
pca_scMerge_unsupervised_cellTypes = scater::plotPCA(sce_scMerge, 
                                                   colour_by = "cellTypes",
                                                   run_args = list(exprs_values = "scMerge_unsupervised"))

pca_scMerge_unsupervised_batch = scater::plotPCA(sce_scMerge, 
                                               colour_by = "batch",
                                               run_args = list(exprs_values = "scMerge_unsupervised"))

grid.arrange(pca_scMerge_unsupervised_cellTypes, pca_scMerge_unsupervised_batch, nrow = 1)
```


# Semi-supervised scMerge

```{r, fig.width=6, fig.height=6}
sce_scMerge <- scMerge(
  sce_combine = sce_scMerge,
  ctl = which(rownames(sce_scMerge) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
  kmeansK = c(6,2),
  replicate_prop = 1,
  WV = sce_scMerge$stage,
  # WV_marker = c("Afp","Alb","Epcam"),
  WV_marker = c("ENSMUSG00000045394","ENSMUSG00000054932","ENSMUSG00000045394"),
  assay_name = "scMerge_semisupervised",
  verbose = TRUE)
```


```{r, fig.width=12, fig.height=6, message=FALSE}
pca_scMerge_semisupervised_cellTypes = scater::plotPCA(sce_scMerge, 
                                                   colour_by = "cellTypes",
                                                   run_args = list(exprs_values = "scMerge_semisupervised"))

pca_scMerge_semisupervised_batch = scater::plotPCA(sce_scMerge, 
                                               colour_by = "batch",
                                               run_args = list(exprs_values = "scMerge_semisupervised"))

grid.arrange(pca_scMerge_semisupervised_cellTypes, pca_scMerge_semisupervised_batch, nrow = 1)
```



<!-- # Extension: Speeding up the calculations -->


<!-- ```{r, eval = FALSE} -->
<!-- system.time( -->
<!--   sce_fast <- scMerge( -->
<!--     sce_combine = sce_combine, -->
<!--     ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG), -->
<!--     kmeansK = c(6,2,3,3), -->
<!--     # exprs = "logcounts", -->
<!--     # hvg_exprs = "counts", -->
<!--     # marker = NULL, -->
<!--     # ruvK = 20, -->
<!--     # cell_type = sce_combine$cellTypes, -->
<!--     # dist = "cor", -->
<!--     replicate_prop = 1, -->
<!--     WV = sce_combine$stage, -->
<!--     # WV_marker = c("Afp","Alb","Epcam"), -->
<!--     WV_marker = c("ENSMUSG00000045394","ENSMUSG00000054932","ENSMUSG00000045394"), -->
<!--     # return_all_RUV = FALSE, -->
<!--     assay_name = "scMerge_fast", -->
<!--     verbose = TRUE, -->
<!--     fast_svd = TRUE, rsvd_prop = 0.1) -->
<!-- ) -->
<!-- ``` -->


# Session Info


```{r, echo = FALSE}
t2 = Sys.time()
```



```{r}
sessionInfo()
t2 - t1
```

