<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Sydney Precision Bioinformatics Group" />

<meta name="date" content="2019-03-05" />

<title>scMerge package</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 64px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h2 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h3 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h4 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h5 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h6 {
  padding-top: 69px;
  margin-top: -69px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SingleCellPlus - HKU Workshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="qc.html">QC Process</a>
</li>
<li>
  <a href="scMerge.html">scMerge</a>
</li>
<li>
  <a href="downstream.html">Downstream Analysis</a>
</li>
<li>
  <a href="acknowledgement.html">acknowledgement</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">scMerge package</h1>
<h4 class="author">Sydney Precision Bioinformatics Group</h4>
<h4 class="date">03/05/2019</h4>

</div>


<p>As discussed in 3rd May 2019, Kevin will work on this part of the workshop.</p>
<p>However, the raw data will be prepared by Yingxin and QC of the data are yet to be done.</p>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>scMerge is a method developed by the Sydney Precision Bioinformatics Group. It aims to merge multiple scRNA-Seq data so that researchers can look for biological signals on data pooled from multiple sources. The key to achieving a good merge is to remove the <strong>noise</strong>.</p>
<p>Here are some plots where the raw data is separated by the data-batches but not cell types.</p>
<p>scMerge corrects data-batch separations by</p>
<ol style="list-style-type: decimal">
<li>identifying cells of identical cell types and pooling these together</li>
<li>identitying data-batch noises using Stably Expressed Genes (SEGs) and removing these from the data.</li>
</ol>
<p>Thus, scMerge can be thought of a two-step process!</p>
</div>
<div id="supervised-scmerge" class="section level1">
<h1>Supervised scMerge</h1>
<p>We will begin with a liver dataset. The data are stored as four separate RDS files (this is similar to a <code>RData</code> file, but only one data can be stored in each RDS file).</p>
<table style="width:100%;">
<colgroup>
<col width="5%" />
<col width="8%" />
<col width="6%" />
<col width="21%" />
<col width="9%" />
<col width="8%" />
<col width="6%" />
<col width="13%" />
<col width="9%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>ID</th>
<th>Author</th>
<th>DOI or URL</th>
<th>Protocol</th>
<th>Organism</th>
<th>Tissue</th>
<th># of cell types</th>
<th># of cells</th>
<th># of batches</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Liver</td>
<td>GSE87795</td>
<td>Su</td>
<td>10.1186/s12864-017-4342-x</td>
<td>SMARTer/C1</td>
<td>Mouse</td>
<td>Liver</td>
<td>6</td>
<td>1236</td>
<td></td>
</tr>
<tr class="even">
<td></td>
<td>GSE90047</td>
<td>Yang</td>
<td>10.1002/hep.29353</td>
<td>Smart-Seq2</td>
<td></td>
<td></td>
<td>2</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>GSE87038</td>
<td>Dong</td>
<td>10.1186/s13059-018-1416-2</td>
<td>STRT-seq</td>
<td></td>
<td></td>
<td>3</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td></td>
<td>GSE96981</td>
<td>Camp</td>
<td>10.1038/nature22796</td>
<td>SMARTer/C1</td>
<td></td>
<td></td>
<td>3</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<pre class="r"><code>su = readRDS(&quot;data/sce_GSE87795.rds&quot;)
yang = readRDS(&quot;data/sce_GSE90047.rds&quot;)
dong = readRDS(&quot;data/sce_GSE87038.rds&quot;)
camp = readRDS(&quot;data/sce_GSE96981.rds&quot;)
sce_list = list(su = su, yang = yang, dong = dong, camp = camp)</code></pre>
<pre class="r"><code>library(scMerge)
library(scater)
library(dplyr)


purrr::map(sce_list, ~ colData(.x) %&gt;% colnames)</code></pre>
<pre><code>## $su
##  [1] &quot;batch&quot;                                     
##  [2] &quot;stage&quot;                                     
##  [3] &quot;cellTypes&quot;                                 
##  [4] &quot;total_features&quot;                            
##  [5] &quot;log10_total_features&quot;                      
##  [6] &quot;total_counts&quot;                              
##  [7] &quot;log10_total_counts&quot;                        
##  [8] &quot;pct_counts_top_50_features&quot;                
##  [9] &quot;pct_counts_top_100_features&quot;               
## [10] &quot;pct_counts_top_200_features&quot;               
## [11] &quot;pct_counts_top_500_features&quot;               
## [12] &quot;total_features_endogenous&quot;                 
## [13] &quot;log10_total_features_endogenous&quot;           
## [14] &quot;total_counts_endogenous&quot;                   
## [15] &quot;log10_total_counts_endogenous&quot;             
## [16] &quot;pct_counts_endogenous&quot;                     
## [17] &quot;pct_counts_top_50_features_endogenous&quot;     
## [18] &quot;pct_counts_top_100_features_endogenous&quot;    
## [19] &quot;pct_counts_top_200_features_endogenous&quot;    
## [20] &quot;pct_counts_top_500_features_endogenous&quot;    
## [21] &quot;total_features_feature_control&quot;            
## [22] &quot;log10_total_features_feature_control&quot;      
## [23] &quot;total_counts_feature_control&quot;              
## [24] &quot;log10_total_counts_feature_control&quot;        
## [25] &quot;pct_counts_feature_control&quot;                
## [26] &quot;pct_counts_top_50_features_feature_control&quot;
## [27] &quot;total_features_ERCC&quot;                       
## [28] &quot;log10_total_features_ERCC&quot;                 
## [29] &quot;total_counts_ERCC&quot;                         
## [30] &quot;log10_total_counts_ERCC&quot;                   
## [31] &quot;pct_counts_ERCC&quot;                           
## [32] &quot;pct_counts_top_50_features_ERCC&quot;           
## [33] &quot;is_cell_control&quot;                           
## 
## $yang
##  [1] &quot;stage&quot;                                     
##  [2] &quot;batch&quot;                                     
##  [3] &quot;cellTypes&quot;                                 
##  [4] &quot;total_features&quot;                            
##  [5] &quot;log10_total_features&quot;                      
##  [6] &quot;total_counts&quot;                              
##  [7] &quot;log10_total_counts&quot;                        
##  [8] &quot;pct_counts_top_50_features&quot;                
##  [9] &quot;pct_counts_top_100_features&quot;               
## [10] &quot;pct_counts_top_200_features&quot;               
## [11] &quot;pct_counts_top_500_features&quot;               
## [12] &quot;total_features_endogenous&quot;                 
## [13] &quot;log10_total_features_endogenous&quot;           
## [14] &quot;total_counts_endogenous&quot;                   
## [15] &quot;log10_total_counts_endogenous&quot;             
## [16] &quot;pct_counts_endogenous&quot;                     
## [17] &quot;pct_counts_top_50_features_endogenous&quot;     
## [18] &quot;pct_counts_top_100_features_endogenous&quot;    
## [19] &quot;pct_counts_top_200_features_endogenous&quot;    
## [20] &quot;pct_counts_top_500_features_endogenous&quot;    
## [21] &quot;total_features_feature_control&quot;            
## [22] &quot;log10_total_features_feature_control&quot;      
## [23] &quot;total_counts_feature_control&quot;              
## [24] &quot;log10_total_counts_feature_control&quot;        
## [25] &quot;pct_counts_feature_control&quot;                
## [26] &quot;pct_counts_top_50_features_feature_control&quot;
## [27] &quot;total_features_ERCC&quot;                       
## [28] &quot;log10_total_features_ERCC&quot;                 
## [29] &quot;total_counts_ERCC&quot;                         
## [30] &quot;log10_total_counts_ERCC&quot;                   
## [31] &quot;pct_counts_ERCC&quot;                           
## [32] &quot;pct_counts_top_50_features_ERCC&quot;           
## [33] &quot;is_cell_control&quot;                           
## 
## $dong
##  [1] &quot;cellTypes1&quot;                  &quot;stage&quot;                      
##  [3] &quot;cellTypes&quot;                   &quot;total_features&quot;             
##  [5] &quot;log10_total_features&quot;        &quot;total_counts&quot;               
##  [7] &quot;log10_total_counts&quot;          &quot;pct_counts_top_50_features&quot; 
##  [9] &quot;pct_counts_top_100_features&quot; &quot;pct_counts_top_200_features&quot;
## [11] &quot;pct_counts_top_500_features&quot; &quot;is_cell_control&quot;            
## 
## $camp
## [1] &quot;cellTypes&quot; &quot;stage&quot;</code></pre>
<pre class="r"><code>sce_combine = scMerge::sce_cbind(sce_list = sce_list, method = &quot;union&quot;, colData_names = c(&quot;cellTypes&quot;, &quot;stage&quot;))</code></pre>
<pre><code>## The assay named &#39;counts&#39; will be used to determine the proportion of zeroes for each batch</code></pre>
<pre class="r"><code>sce_combine</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 23110 1248 
## metadata(0):
## assays(2): counts logcounts
## rownames(23110): ENSMUSG00000102331 ENSMUSG00000025902 ...
##   __ambiguous __alignment_not_unique
## rowData names(0):
## colnames(1248): E11.5_C07 E11.5_C44 ... F5A_E16.5 F6A_E16.5
## colData names(3): cellTypes stage batch
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<pre class="r"><code>data(&quot;segList_ensemblGeneID&quot;, package = &quot;scMerge&quot;)

sce_combine &lt;- scMerge(sce_combine = sce_combine, 
                       ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
                       kmeansK = c(6,2,3,3),
                       exprs = &quot;logcounts&quot;,
                       hvg_exprs = &quot;counts&quot;,
                       marker = NULL,
                       ruvK = 20,
                       cell_type = NULL,
                       dist = &quot;cor&quot;,
                       replicate_prop = 1,
                       WV = sce_combine$stage,
                       # WV_marker = c(&quot;Afp&quot;,&quot;Alb&quot;,&quot;Epcam&quot;),
                       WV_marker = c(&quot;ENSMUSG00000045394&quot;,&quot;ENSMUSG00000054932&quot;,&quot;ENSMUSG00000045394&quot;),
                       return_all_RUV = FALSE,
                       assay_name = &quot;scMerge_scSEG&quot;,
                       verbose = TRUE,
                       fast_svd = FALSE)</code></pre>
<pre><code>## Step 1: Computation will run in serial</code></pre>
<pre><code>## Performing unsupervised scMerge with: 
##  1. No cell type information 
##  2. Cell type indices not revelant here 
##  3. Mutual nearest neighbour matching 
##  4. No supplied marker and no supplied marker_list for MNN clustering 
##     Finding Highly Variable Genes for clustering 
##     4515 HVG were found 
##  5. PCA and Kmeans clustering will be performed on each batch 
##  6. Create Mutual Nearest Clusters. Preview cells-cell_type matching output matrix:</code></pre>
<pre><code>##   group batch cluster
## 1     1     1       2
## 2     1     2       1
## 3     1     3       2
## 4     4     1       5
## 5     4     3       1
## 6     2     1       6
##  7. Performing semi-supervised scMerge with wanted variation 
## Dimension of the replicates mapping matrix: 
## [1] 1248   19</code></pre>
<pre><code>## Step 2: Performing RUV normalisation. This will take minutes to hours.</code></pre>
<pre><code>## scMerge complete!</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<pre class="r"><code>scater::plotTSNE(sce_combine, 
                colour_by = &#39;cellTypes&#39;, shape = &#39;batch&#39;,
                run_args = list(exprs_values = &#39;logcounts&#39;))</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<pre class="r"><code>scater::plotTSNE(sce_combine, 
                colour_by = &#39;cellTypes&#39;, shape = &#39;batch&#39;,
                run_args = list(exprs_values = &#39;scMerge_scSEG&#39;))</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-5-2.png" width="672" /></p>
</div>
<div id="unsupervised-scmerge" class="section level1">
<h1>Unsupervised scMerge</h1>
</div>
<div id="extension-semi-supervised-scmerge" class="section level1">
<h1>Extension: semi-supervised scMerge</h1>
</div>
<div id="session-info" class="section level1">
<h1>Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] dplyr_0.8.1                 scater_1.12.1              
##  [3] ggplot2_3.1.1               SingleCellExperiment_1.6.0 
##  [5] SummarizedExperiment_1.14.0 DelayedArray_0.10.0        
##  [7] BiocParallel_1.18.0         matrixStats_0.54.0         
##  [9] Biobase_2.44.0              GenomicRanges_1.36.0       
## [11] GenomeInfoDb_1.20.0         IRanges_2.18.0             
## [13] S4Vectors_0.22.0            BiocGenerics_0.30.0        
## [15] scMerge_1.0.0               rmarkdown_1.12             
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-140             bitops_1.0-6            
##  [3] RColorBrewer_1.1-2       numDeriv_2016.8-1       
##  [5] tools_3.6.0              backports_1.1.4         
##  [7] R6_2.4.0                 irlba_2.3.3             
##  [9] rpart_4.1-15             KernSmooth_2.23-15      
## [11] vipor_0.4.5              uwot_0.1.3              
## [13] mgcv_1.8-28              Hmisc_4.2-0             
## [15] lazyeval_0.2.2           colorspace_1.4-1        
## [17] nnet_7.3-12              withr_2.1.2             
## [19] tidyselect_0.2.5         gridExtra_2.3           
## [21] compiler_3.6.0           BiocNeighbors_1.2.0     
## [23] htmlTable_1.13.1         labeling_0.3            
## [25] caTools_1.17.1.2         scales_1.0.0            
## [27] checkmate_1.9.3          proxy_0.4-23            
## [29] stringr_1.4.0            digest_0.6.19           
## [31] foreign_0.8-71           XVector_0.24.0          
## [33] base64enc_0.1-3          pkgconfig_2.0.2         
## [35] htmltools_0.3.6          ruv_0.9.7               
## [37] bbmle_1.0.20             htmlwidgets_1.3         
## [39] rlang_0.3.4              rstudioapi_0.10         
## [41] FNN_1.1.3                DelayedMatrixStats_1.6.0
## [43] gtools_3.8.1             acepack_1.4.1           
## [45] RCurl_1.95-4.12          magrittr_1.5            
## [47] BiocSingular_1.0.0       GenomeInfoDbData_1.2.1  
## [49] Formula_1.2-3            Matrix_1.2-17           
## [51] Rcpp_1.0.1               ggbeeswarm_0.6.0        
## [53] munsell_0.5.0            viridis_0.5.1           
## [55] stringi_1.4.3            yaml_2.2.0              
## [57] zlibbioc_1.30.0          Rtsne_0.15              
## [59] gplots_3.0.1.1           plyr_1.8.4              
## [61] grid_3.6.0               gdata_2.18.0            
## [63] M3Drop_1.10.0            reldist_1.6-6           
## [65] crayon_1.3.4             lattice_0.20-38         
## [67] cowplot_0.9.4            splines_3.6.0           
## [69] knitr_1.23               pillar_1.4.0            
## [71] igraph_1.2.4.1           codetools_0.2-16        
## [73] glue_1.3.1               packrat_0.5.0           
## [75] evaluate_0.13            latticeExtra_0.6-28     
## [77] RcppParallel_4.4.2       data.table_1.12.2       
## [79] foreach_1.4.4            gtable_0.3.0            
## [81] purrr_0.3.2              assertthat_0.2.1        
## [83] xfun_0.7                 rsvd_1.0.0              
## [85] RcppEigen_0.3.3.5.0      RSpectra_0.14-0         
## [87] survival_2.44-1.1        viridisLite_0.3.0       
## [89] tibble_2.1.1             iterators_1.0.10        
## [91] beeswarm_0.2.3           cluster_2.0.9           
## [93] statmod_1.4.30</code></pre>
<pre class="r"><code>t2 - t1</code></pre>
<pre><code>## Time difference of 1.112507 mins</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
