<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Sydney Precision Bioinformatics Group" />

<meta name="date" content="2019-03-05" />

<title>scMerge package</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 64px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h2 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h3 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h4 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h5 {
  padding-top: 69px;
  margin-top: -69px;
}
.section h6 {
  padding-top: 69px;
  margin-top: -69px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>



<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">SingleCellPlus - HKU Workshop</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="qc.html">QC Process</a>
</li>
<li>
  <a href="scMerge.html">scMerge</a>
</li>
<li>
  <a href="downstream.html">Downstream Analysis</a>
</li>
<li>
  <a href="acknowledgement.html">Acknowledgement</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">scMerge package</h1>
<h4 class="author">Sydney Precision Bioinformatics Group</h4>
<h4 class="date">03/05/2019</h4>

</div>


<div id="jeans-comments" class="section level1">
<h1><span class="header-section-number">1</span> Jean’s comments</h1>
<ol style="list-style-type: decimal">
<li><code>str</code> and <code>fct</code> manipulations as blackbox.</li>
<li>Change “batch” into “data sources”/“datasets”</li>
<li>Separate plotly plots into two</li>
<li>Only keep Su and Yang.</li>
<li>Dropping E10.5 and E16.5</li>
</ol>
</div>
<div id="introduction" class="section level1">
<h1><span class="header-section-number">2</span> Introduction</h1>
<p>scMerge is a method developed by the Sydney Precision Bioinformatics Group. It aims to merge multiple scRNA-Seq data so that researchers can look for biological signals on data pooled from multiple sources. The key to achieving a good merge is to remove the <strong>noise</strong>.</p>
<p>Here are some plots where the raw data is separated by the data-batches but not cell types.</p>
<p>scMerge corrects data-batch separations by</p>
<ol style="list-style-type: decimal">
<li>identifying cells of identical cell types and pooling these together</li>
<li>identitying data-batch noises using Stably Expressed Genes (SEGs) and removing these from the data.</li>
</ol>
<p>Thus, scMerge can be thought of a two-step process!</p>
</div>
<div id="supervised-scmerge" class="section level1">
<h1><span class="header-section-number">3</span> Supervised scMerge</h1>
<p>We will begin with a liver dataset. The data are stored as four separate RDS files (this is similar to a <code>RData</code> file, but only one data can be stored in each RDS file).</p>
<table style="width:100%;">
<colgroup>
<col width="5%" />
<col width="8%" />
<col width="6%" />
<col width="21%" />
<col width="9%" />
<col width="8%" />
<col width="6%" />
<col width="13%" />
<col width="9%" />
<col width="11%" />
</colgroup>
<thead>
<tr class="header">
<th>Name</th>
<th>ID</th>
<th>Author</th>
<th>DOI or URL</th>
<th>Protocol</th>
<th>Organism</th>
<th>Tissue</th>
<th># of cell types</th>
<th># of cells</th>
<th># of batches</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Liver</td>
<td>GSE87795</td>
<td>Su</td>
<td>10.1186/s12864-017-4342-x</td>
<td>SMARTer/C1</td>
<td>Mouse</td>
<td>Liver</td>
<td>6</td>
<td>1236</td>
<td>6</td>
</tr>
<tr class="even">
<td></td>
<td>GSE90047</td>
<td>Yang</td>
<td>10.1002/hep.29353</td>
<td>Smart-Seq2</td>
<td></td>
<td></td>
<td>2</td>
<td></td>
<td>2</td>
</tr>
<tr class="odd">
<td></td>
<td>GSE87038</td>
<td>Dong</td>
<td>10.1186/s13059-018-1416-2</td>
<td>STRT-seq</td>
<td></td>
<td></td>
<td>4</td>
<td></td>
<td>3</td>
</tr>
<tr class="even">
<td></td>
<td>GSE96981</td>
<td>Camp</td>
<td>10.1038/nature22796</td>
<td>SMARTer/C1</td>
<td></td>
<td></td>
<td>3</td>
<td></td>
<td>3</td>
</tr>
</tbody>
</table>
<pre class="r"><code>su = readRDS(&quot;data/sce_GSE87795.rds&quot;)
yang = readRDS(&quot;data/sce_GSE90047.rds&quot;)
# dong = readRDS(&quot;data/sce_GSE87038.rds&quot;)
# camp = readRDS(&quot;data/sce_GSE96981.rds&quot;)
sce_list = list(
  su = su, 
  yang = yang
  # dong = dong, 
  # camp = camp
)</code></pre>
<pre class="r"><code>library(scMerge)
library(scater)
library(dplyr)
library(gridExtra)
library(plotly)
library(stringr)
library(forcats)

library(tibble)
library(dplyr)

theme_set(theme_classic(16)
            # theme(legend.position = &quot;bottom&quot;)
          )


# purrr::map(sce_list, ~ colData(.x) %&gt;% colnames)

sce_combine = scMerge::sce_cbind(sce_list = sce_list, 
                                 method = &quot;union&quot;, 
                                 colData_names = c(&quot;cellTypes&quot;, &quot;stage&quot;))


cat(&quot;Please copy and paste this&quot;)</code></pre>
<pre><code>## Please copy and paste this</code></pre>
<pre class="r"><code># sce_combine = sce_combine[,stringr::str_detect(sce_combine$cellTypes, &quot;cholangiocyte|Epithelial Cell|hepatoblast|Hepatoblast|Mesenchymal Cell|Endothelial Cell|Hematopoietic|Immune cell|Stellate Cell&quot;)]

colData(sce_combine)$cellTypes = forcats::fct_recode(
  colData(sce_combine)$cellTypes, 
  Hepatoblast = &quot;hepatoblast/hepatocyte&quot;
  # Hepatoblast = &quot;Hepatoblasts&quot;
) %&gt;% 
  droplevels()

sce_combine = sce_combine[rowSums(SingleCellExperiment::counts(sce_combine)) != 0,
                          colSums(SingleCellExperiment::counts(sce_combine)) != 0]</code></pre>
<pre class="r"><code>table(sce_combine$cellTypes, sce_combine$batch)</code></pre>
<pre><code>##                   
##                    batch 1 batch 2
##   Endothelial Cell      52       0
##   Erythrocyte          132       0
##   Hepatoblast           49     345
##   Macrophage            53       0
##   Megakaryocyte         17       0
##   Mesenchymal Cell      86       0
##   cholangiocyte          0     103</code></pre>
<pre class="r"><code>table(sce_combine$cellTypes, sce_combine$stage)</code></pre>
<pre><code>##                   
##                    E11.5 E12.5 E13.5 E14.5 E16.5 E10.5 E15.5 E17.5
##   Endothelial Cell     8    12    12    13     7     0     0     0
##   Erythrocyte          1    34    30    43    24     0     0     0
##   Hepatoblast         70    45    66    70    12    54    41    36
##   Macrophage           3     6     9    12    23     0     0     0
##   Megakaryocyte        1     7     5     2     2     0     0     0
##   Mesenchymal Cell    43    25     4    10     4     0     0     0
##   cholangiocyte        2     3    10    17     0     0    36    35</code></pre>
<pre class="r"><code>table(sce_combine$batch, sce_combine$stage)</code></pre>
<pre><code>##          
##           E11.5 E12.5 E13.5 E14.5 E16.5 E10.5 E15.5 E17.5
##   batch 1    58    91    71    97    72     0     0     0
##   batch 2    70    41    65    70     0    54    77    71</code></pre>
<pre class="r"><code>cell_data = colData(sce_combine) %&gt;% 
  as.data.frame() %&gt;% 
  group_by(cellTypes, stage, batch) %&gt;%
  dplyr::summarise(n = n()) %&gt;% 
  ungroup() %&gt;% 
  tidyr::complete(cellTypes, stage, batch, fill = list(n = 0)) %&gt;%
  dplyr::mutate(
    stage = factor(stage, levels = sort(unique(levels(stage))))
    )

cell_data %&gt;% 
  ggplot(aes(x = stage, y = cellTypes, fill = n, label = n)) +
  geom_tile() +
  geom_text() +
  facet_wrap(~batch) +
  scale_fill_distiller(palette = &quot;Reds&quot;, direction = 1) +
  labs(title = &quot;Number of cells split by batch, celltypes and stage&quot;)</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-4-1.png" width="1152" /></p>
<pre class="r"><code>pca_logcounts_cellTypes = scater::plotPCA(sce_combine, 
                                          colour_by = &quot;cellTypes&quot;,
                                          run_args = list(exprs_values = &quot;logcounts&quot;)) +
  scale_fill_brewer(palette = &quot;Set1&quot;)

pca_logcounts_batch = scater::plotPCA(sce_combine, 
                                      colour_by = &quot;batch&quot;,
                                      run_args = list(exprs_values = &quot;logcounts&quot;)) +
  scale_fill_brewer(palette = &quot;Dark2&quot;)

grid.arrange(pca_logcounts_cellTypes, pca_logcounts_batch, nrow = 1)</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-4-2.png" width="1152" /></p>
<pre class="r"><code>data(&quot;segList_ensemblGeneID&quot;, package = &quot;scMerge&quot;)

sce_scMerge &lt;- scMerge(
  sce_combine = sce_combine,
  ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
  cell_type = sce_combine$cellTypes,
  replicate_prop = 1,
  assay_name = &quot;scMerge_supervised&quot;,
  verbose = TRUE)</code></pre>
<pre><code>## Step 1: Computation will run in serial</code></pre>
<pre><code>## Performing supervised scMerge with: 
##  1. Cell type information 
##  2. No cell type indices 
##  3. No mutual nearest neighbour clustering 
## Dimension of the replicates mapping matrix: 
## [1] 837   7</code></pre>
<pre><code>## Step 2: Performing RUV normalisation. This will take minutes to hours.</code></pre>
<pre><code>## scMerge complete!</code></pre>
<pre class="r"><code>sce_scMerge</code></pre>
<pre><code>## class: SingleCellExperiment 
## dim: 23801 837 
## metadata(5): ruvK ruvK_optimal scRep_res timeReplicates timeRuv
## assays(3): counts logcounts scMerge_supervised
## rownames(23801): ENSMUSG00000104017 ENSMUSG00000102331 ...
##   ERCC-00158 ERCC-00164
## rowData names(0):
## colnames(837): E11.5_C07 E11.5_C44 ... E17.5E_4_16 E17.5E_4_16.1
## colData names(3): cellTypes stage batch
## reducedDimNames(0):
## spikeNames(0):</code></pre>
<pre class="r"><code>pca_scMerge_supervised_cellTypes = scater::plotPCA(sce_scMerge, 
                                                   colour_by = &quot;cellTypes&quot;,
                                                   run_args = list(exprs_values = &quot;scMerge_supervised&quot;)) +
  scale_fill_brewer(palette = &quot;Set1&quot;)

pca_scMerge_supervised_batch = scater::plotPCA(sce_scMerge, 
                                               colour_by = &quot;batch&quot;,
                                               run_args = list(exprs_values = &quot;scMerge_supervised&quot;)) +
  scale_fill_brewer(palette = &quot;Dark2&quot;)

grid.arrange(pca_scMerge_supervised_cellTypes, pca_scMerge_supervised_batch, nrow = 1)</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-6-1.png" width="1152" /></p>
<pre class="r"><code>supervised_cluster = metadata(sce_scMerge)[[&quot;scRep_res&quot;]] %&gt;% 
  apply(1, which.max)</code></pre>
</div>
<div id="unsupervised-scmerge" class="section level1">
<h1><span class="header-section-number">4</span> Unsupervised scMerge</h1>
<pre class="r"><code>sce_scMerge &lt;- scMerge(
  sce_combine = sce_combine,
  ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
  kmeansK = c(6, 2),
  replicate_prop = 1,
  assay_name = &quot;scMerge_unsupervised&quot;,
  verbose = TRUE)</code></pre>
<pre><code>## Step 1: Computation will run in serial</code></pre>
<pre><code>## Performing unsupervised scMerge with: 
##  1. No cell type information 
##  2. Cell type indices not revelant here 
##  3. Mutual nearest neighbour matching 
##  4. No supplied marker and no supplied marker_list for MNN clustering 
##     Finding Highly Variable Genes for clustering 
##     3650 HVG were found 
##  5. PCA and Kmeans clustering will be performed on each batch 
##  6. Create Mutual Nearest Clusters. Preview cells-cell_type matching output matrix:</code></pre>
<pre><code>##   group batch cluster
## 1     1     1       5
## 2     2     2       2
## Dimension of the replicates mapping matrix: 
## [1] 837   8</code></pre>
<pre><code>## Step 2: Performing RUV normalisation. This will take minutes to hours.</code></pre>
<pre><code>## scMerge complete!</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-7-1.png" width="576" /></p>
<pre class="r"><code>pca_scMerge_unsupervised_cellTypes = scater::plotPCA(sce_scMerge, 
                                                   colour_by = &quot;cellTypes&quot;,
                                                   run_args = list(exprs_values = &quot;scMerge_unsupervised&quot;)) +
  scale_fill_brewer(palette = &quot;Set1&quot;)

pca_scMerge_unsupervised_batch = scater::plotPCA(sce_scMerge, 
                                               colour_by = &quot;batch&quot;,
                                               run_args = list(exprs_values = &quot;scMerge_unsupervised&quot;)) +
  scale_fill_brewer(palette = &quot;Dark2&quot;)

grid.arrange(pca_scMerge_unsupervised_cellTypes, pca_scMerge_unsupervised_batch, nrow = 1)</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-8-1.png" width="1152" /></p>
<pre class="r"><code>unsupervised_cluster = metadata(sce_scMerge)[[&quot;scRep_res&quot;]] %&gt;% 
  apply(1, which.max)

sce_pca = scater::runPCA(sce_scMerge, exprs_values = &quot;scMerge_unsupervised&quot;)
pca_mat = sce_pca@reducedDims@listData$PCA
colnames(pca_mat) = c(&quot;pca1&quot;, &quot;pca2&quot;)
pca_df = bind_cols(data.frame(pca_mat), as.data.frame(colData(sce_pca))) %&gt;% 
  mutate(
    supervised_cluster = as.factor(supervised_cluster), 
    unsupervised_cluster = as.factor(unsupervised_cluster))

pca_df_summ = pca_df %&gt;% 
  group_by(supervised_cluster, cellTypes, unsupervised_cluster) %&gt;% 
  dplyr::tally()

pca_df_summ</code></pre>
<pre><code>## # A tibble: 18 x 4
## # Groups:   supervised_cluster, cellTypes [7]
##    supervised_cluster cellTypes        unsupervised_cluster     n
##    &lt;fct&gt;              &lt;fct&gt;            &lt;fct&gt;                &lt;int&gt;
##  1 1                  Endothelial Cell 3                       48
##  2 1                  Endothelial Cell 5                        4
##  3 2                  Erythrocyte      3                        2
##  4 2                  Erythrocyte      4                       28
##  5 2                  Erythrocyte      6                      102
##  6 3                  Hepatoblast      1                       48
##  7 3                  Hepatoblast      2                      345
##  8 3                  Hepatoblast      5                        1
##  9 4                  Macrophage       3                        1
## 10 4                  Macrophage       4                       18
## 11 4                  Macrophage       5                        4
## 12 4                  Macrophage       6                        1
## 13 4                  Macrophage       7                       29
## 14 5                  Megakaryocyte    4                       17
## 15 6                  Mesenchymal Cell 5                       85
## 16 6                  Mesenchymal Cell 6                        1
## 17 7                  cholangiocyte    2                       10
## 18 7                  cholangiocyte    8                       93</code></pre>
<pre class="r"><code>pca_df_summ %&gt;% 
  ggplot(aes(x = unsupervised_cluster, 
             y = cellTypes,
             fill = n, 
             label = n)) +
  geom_tile() +
  geom_text()</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-8-2.png" width="1152" /></p>
</div>
<div id="semi-supervised-scmerge" class="section level1">
<h1><span class="header-section-number">5</span> Semi-supervised scMerge</h1>
<pre class="r"><code>sce_scMerge &lt;- scMerge(
  sce_combine = sce_combine,
  ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG),
  kmeansK = c(6,2),
  replicate_prop = 1,
  WV = sce_scMerge$stage,
  # WV_marker = c(&quot;Afp&quot;,&quot;Alb&quot;,&quot;Epcam&quot;),
  WV_marker = c(&quot;ENSMUSG00000045394&quot;,&quot;ENSMUSG00000054932&quot;,&quot;ENSMUSG00000045394&quot;),
  assay_name = &quot;scMerge_semisupervised&quot;,
  verbose = TRUE)</code></pre>
<pre><code>## Step 1: Computation will run in serial</code></pre>
<pre><code>## Performing unsupervised scMerge with: 
##  1. No cell type information 
##  2. Cell type indices not revelant here 
##  3. Mutual nearest neighbour matching 
##  4. No supplied marker and no supplied marker_list for MNN clustering 
##     Finding Highly Variable Genes for clustering 
##     3650 HVG were found 
##  5. PCA and Kmeans clustering will be performed on each batch 
##  6. Create Mutual Nearest Clusters. Preview cells-cell_type matching output matrix:</code></pre>
<pre><code>##   group batch cluster
## 1     1     1       5
## 2     2     2       1
##  7. Performing semi-supervised scMerge with wanted variation 
## Dimension of the replicates mapping matrix: 
## [1] 837  22</code></pre>
<pre><code>## Step 2: Performing RUV normalisation. This will take minutes to hours.</code></pre>
<pre><code>## scMerge complete!</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-9-1.png" width="576" /></p>
<pre class="r"><code>pca_scMerge_semisupervised_cellTypes = scater::plotPCA(sce_scMerge, 
                                                   colour_by = &quot;cellTypes&quot;,
                                                   run_args = list(exprs_values = &quot;scMerge_semisupervised&quot;))

pca_scMerge_semisupervised_batch = scater::plotPCA(sce_scMerge, 
                                               colour_by = &quot;batch&quot;,
                                               run_args = list(exprs_values = &quot;scMerge_semisupervised&quot;))

grid.arrange(pca_scMerge_semisupervised_cellTypes, pca_scMerge_semisupervised_batch, nrow = 1)</code></pre>
<p><img src="scMerge_files/figure-html/unnamed-chunk-10-1.png" width="1152" /></p>
<!-- ```{r} -->
<!-- scater::plotExpression(sce_scMerge[,sce_scMerge$cellTypes == "Hepatoblast"], -->
<!--                        features = c("ENSMUSG00000045394","ENSMUSG00000054932","ENSMUSG00000045394"),  -->
<!--                        exprs_values = "scMerge_supervised",  -->
<!--                        colour_by = "batch") + -->
<!--   labs(title = "Expression of the WV marker genes for Hepatoblast only in supervied-scMerge") -->
<!-- scater::plotExpression(sce_scMerge[,sce_scMerge$cellTypes == "Hepatoblast"], -->
<!--                        features = c("ENSMUSG00000045394","ENSMUSG00000054932","ENSMUSG00000045394"),  -->
<!--                        exprs_values = "scMerge_unsupervised",  -->
<!--                        colour_by = "batch") + -->
<!--   labs(title = "Expression of the WV marker genes for Hepatoblast only in unsupervied-scMerge") -->
<!-- scater::plotExpression(sce_scMerge[,sce_scMerge$cellTypes == "Hepatoblast"], -->
<!--                        features = c("ENSMUSG00000045394","ENSMUSG00000054932","ENSMUSG00000045394"),  -->
<!--                        exprs_values = "scMerge_semisupervised",  -->
<!--                        colour_by = "batch") + -->
<!--   labs(title = "Expression of the WV marker genes for Hepatoblast only in semisupervied-scMerge") -->
<!-- ``` -->
<!-- # Extension: Speeding up the calculations -->
<!-- ```{r, eval = FALSE} -->
<!-- system.time( -->
<!--   sce_fast <- scMerge( -->
<!--     sce_combine = sce_combine, -->
<!--     ctl = which(rownames(sce_combine) %in% segList_ensemblGeneID$mouse$mouse_scSEG), -->
<!--     kmeansK = c(6,2,3,3), -->
<!--     # exprs = "logcounts", -->
<!--     # hvg_exprs = "counts", -->
<!--     # marker = NULL, -->
<!--     # ruvK = 20, -->
<!--     # cell_type = sce_combine$cellTypes, -->
<!--     # dist = "cor", -->
<!--     replicate_prop = 1, -->
<!--     WV = sce_combine$stage, -->
<!--     # WV_marker = c("Afp","Alb","Epcam"), -->
<!--     WV_marker = c("ENSMUSG00000045394","ENSMUSG00000054932","ENSMUSG00000045394"), -->
<!--     # return_all_RUV = FALSE, -->
<!--     assay_name = "scMerge_fast", -->
<!--     verbose = TRUE, -->
<!--     fast_svd = TRUE, rsvd_prop = 0.1) -->
<!-- ) -->
<!-- ``` -->
</div>
<div id="session-info" class="section level1">
<h1><span class="header-section-number">6</span> Session Info</h1>
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>## R version 3.6.0 (2019-04-26)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_AU.UTF-8/en_AU.UTF-8/en_AU.UTF-8/C/en_AU.UTF-8/en_AU.UTF-8
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] tibble_2.1.2                forcats_0.4.0              
##  [3] stringr_1.4.0               plotly_4.9.0               
##  [5] gridExtra_2.3               dplyr_0.8.1                
##  [7] scater_1.12.2               ggplot2_3.1.1              
##  [9] SingleCellExperiment_1.6.0  SummarizedExperiment_1.14.0
## [11] DelayedArray_0.10.0         BiocParallel_1.18.0        
## [13] matrixStats_0.54.0          Biobase_2.44.0             
## [15] GenomicRanges_1.36.0        GenomeInfoDb_1.20.0        
## [17] IRanges_2.18.1              S4Vectors_0.22.0           
## [19] BiocGenerics_0.30.0         scMerge_1.1.1              
## 
## loaded via a namespace (and not attached):
##  [1] nlme_3.1-140             bitops_1.0-6            
##  [3] httr_1.4.0               RColorBrewer_1.1-2      
##  [5] numDeriv_2016.8-1        tools_3.6.0             
##  [7] backports_1.1.4          utf8_1.1.4              
##  [9] R6_2.4.0                 irlba_2.3.3             
## [11] vipor_0.4.5              rpart_4.1-15            
## [13] KernSmooth_2.23-15       Hmisc_4.2-0             
## [15] lazyeval_0.2.2           mgcv_1.8-28             
## [17] colorspace_1.4-1         nnet_7.3-12             
## [19] withr_2.1.2              tidyselect_0.2.5        
## [21] compiler_3.6.0           cli_1.1.0               
## [23] BiocNeighbors_1.2.0      htmlTable_1.13.1        
## [25] labeling_0.3             caTools_1.17.1.2        
## [27] scales_1.0.0             checkmate_1.9.3         
## [29] proxy_0.4-23             digest_0.6.19           
## [31] foreign_0.8-71           rmarkdown_1.13          
## [33] XVector_0.24.0           base64enc_0.1-3         
## [35] pkgconfig_2.0.2          htmltools_0.3.6         
## [37] ruv_0.9.7                bbmle_1.0.20            
## [39] htmlwidgets_1.3          rlang_0.3.4             
## [41] rstudioapi_0.10          DelayedMatrixStats_1.6.0
## [43] jsonlite_1.6             gtools_3.8.1            
## [45] acepack_1.4.1            RCurl_1.95-4.12         
## [47] magrittr_1.5             BiocSingular_1.0.0      
## [49] GenomeInfoDbData_1.2.1   Formula_1.2-3           
## [51] Matrix_1.2-17            fansi_0.4.0             
## [53] ggbeeswarm_0.6.0         Rcpp_1.0.1              
## [55] munsell_0.5.0            viridis_0.5.1           
## [57] stringi_1.4.3            yaml_2.2.0              
## [59] zlibbioc_1.30.0          gplots_3.0.1.1          
## [61] plyr_1.8.4               grid_3.6.0              
## [63] gdata_2.18.0             M3Drop_1.10.0           
## [65] reldist_1.6-6            crayon_1.3.4            
## [67] lattice_0.20-38          cowplot_0.9.4           
## [69] splines_3.6.0            zeallot_0.1.0           
## [71] knitr_1.23               pillar_1.4.1            
## [73] igraph_1.2.4.1           codetools_0.2-16        
## [75] glue_1.3.1               evaluate_0.14           
## [77] latticeExtra_0.6-28      data.table_1.12.2       
## [79] vctrs_0.1.0              foreach_1.4.4           
## [81] tidyr_0.8.3              gtable_0.3.0            
## [83] purrr_0.3.2              assertthat_0.2.1        
## [85] xfun_0.7                 rsvd_1.0.1              
## [87] RcppEigen_0.3.3.5.0      viridisLite_0.3.0       
## [89] survival_2.44-1.1        iterators_1.0.10        
## [91] beeswarm_0.2.3           cluster_2.0.9           
## [93] statmod_1.4.32</code></pre>
<pre class="r"><code>t2 - t1</code></pre>
<pre><code>## Time difference of 1.355624 mins</code></pre>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
